### PR TITLE
Support after_commit hooks in transactional fixtures

### LINKED ISSUE #16831: Add test_with_transaction to easily test transaction callbacks etc
TIL that you can actually easily transaction callbacks like after_commit, but it's a bit cumbersome. This is what you have to do today:

```ruby
  uses_transaction :test_completing_todo_will_cause_a_chat_relay
  test 'completing todo will cause a chat relay' do
    recordings(:design_todo).complete
    assert last_relayed_recordable(:anniversary).completed?
  end
```

I'd like to expose this as a more user-friendly and documented feature:

```ruby
  test_with_transaction 'completing todo will cause a chat relay' do
    recordings(:design_todo).complete
    assert last_relayed_recordable(:anniversary).completed?
  end
```

### PR DESCRIPTION
**Pull request by brainopia**
Fixes #16831

@jeremy @dhh Have you had something like that in mind?

### CODE REVIEW

**Comment by brainopia** [comment]
This PR changes behavior of transactions inside fixtures. after_commit hooks will be ran for nested transactions if their parent transaction has an `uncommittable` setting. A fixture transaction will always have this setting enabled.

**Comment by brainopia** [comment]
I haven't added any new tests since I've removed `self.use_transactional_fixtures = false` from transactional callback tests to illustrate that they can pass from now on without disabling it.

File: activerecord/test/cases/transaction_callbacks_test.rb [hunk 1]
```diff
@@ -200,21 +199,20 @@ def test_only_call_after_rollback_on_create_after_transaction_rollsback_for_new_
...
+    saved = @first.transaction {
+      tx = @first.class.connection.transaction_manager.current_transaction
+      def tx.commit
+        raise
       end
 
-      @first.after_commit_block{|r| r.history << :after_commit}
-      @first.after_rollback_block{|r| r.history << :after_rollback}
+      @first.save
+    } rescue nil
```
Comment (jeremy) [inline activerecord/test/cases/transaction_callbacks_test.rb]: Rather than `rescue nil` + `assert !saved` you can do
```ruby
assert_raises RuntimeError do
  @first.transaction do
    …
```

File: activerecord/test/cases/transaction_callbacks_test.rb [hunk 2]
```diff
@@ -200,21 +199,20 @@ def test_only_call_after_rollback_on_create_after_transaction_rollsback_for_new_
   end
 
   def test_call_after_rollback_when_commit_fails
-    @first.class.connection.singleton_class.send(:alias_method, :real_method_commit_db_transaction, :commit_db_transaction)
-    begin
-      @first.class.connection.singleton_class.class_eval do
-        def commit_db_transaction; raise "boom!"; end
+    @first.after_commit_block{|r| r.history << :after_commit}
+    @first.after_rollback_block{|r| r.history << :after_rollback}
```
Comment (jeremy) [inline activerecord/test/cases/transaction_callbacks_test.rb]: Code style: separate braces and block args with spaces: `foo { |arg| … }` vs `foo{|arg| …}`

Comment (brainopia) [inline activerecord/test/cases/transaction_callbacks_test.rb]: I'll fix it, I haven't touched these lines in test.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 3]
```diff
@@ -52,6 +52,7 @@ def initialize(connection, options)
         @state = TransactionState.new
         @records = []
         @joinable = options.fetch(:joinable, true)
+        @uncommittable = options[:uncommittable]
```
Comment (jeremy) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: `options.fetch(:uncommittable, false)` to give it a clear default value vs falling back to `nil`

File: activerecord/test/cases/transaction_callbacks_test.rb (hunk 1 above)
Comment (brainopia) [inline activerecord/test/cases/transaction_callbacks_test.rb]: Here is the same, my bad, haven't paid enough attention, just upgraded the old test keeping its `rescue nil; assert !saved` stuff. I'll fix it

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 3 above)
Comment (brainopia) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: If you prefer false to nil in this case, sure I'll change that.

**Comment by brainopia** [comment]
@jeremy Thank you, I've reflected your notes :smile:

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 4]
```diff
@@ -118,7 +120,11 @@ def commit
         connection.release_savepoint(savepoint_name)
         super
         parent = connection.transaction_manager.current_transaction
-        records.each { |r| parent.add_record(r) }
+        if parent.uncommittable?
+          commit_records
```
Comment (jeremy) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: We want to communicate that the app has a logical commit boundary for AR instances that may differ from the physical db commit boundary. "parent.uncommittable?" reads as "parent db txn cannot be committed" when we want to know "should this txn commit its AR instances?" Could rename to something like `participates_in_records_transaction` which defaults to `true`.

Also: should this class even be responsible for figuring out whether it should commit records? Reaching in to the txn manager to check state is a red flag, not to mention that it relies on txn manager having already popped this txn off its stack leaving the parent as the current txn. Perhaps this should be the TransactionManager's duty in the first place:
```ruby
def commit_transaction
  transaction = @stack.pop
  transaction.commit
  transaction.commit_records if !current_transaction.participates_in_records_transaction?
end
```

Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: @jeremy you are 100% right, https://github.com/rails/rails/pull/18426 , I missed that part once I was doing the refactors, see more explanation on my PR.
This will also make the implementation of a TestTransaction easier, as the actual txn should not need to care of copying records to the parent..

Comment (brainopia) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: @arthurnn Thanks for helping out. I'll rebase this PR on your commit :+1: 

@jeremy I've pushed a new version, what do you think about `commit_transaction?` name? Or do you prefer `participates_in_records_transaction?`?

**Comment by arthurnn** [comment]
@brainopia Thanks a lot for this PR, I will look into it this weekend, and add some feedback here.

**Comment by brainopia** [comment]
@arthurnn :bow:

**Comment by arthurnn** [comment]
@brainopia Again, thanks a lot for the PR, however we cannot merge this, which actually has incorrect behaviour. In your PR, all savepoints transactions will run `after_commit` callbacks, and not only under tests that have `use_transactional_fixtures = true` .
On rails 4.2 I added an extra layer (TransactionManager) to solve this type of problems, so we can have custom Transaction managers that handle the transactions in a different form. The problem about testing having to call `after_commit` is a good use case for this. Also that give us more flexibility to implement any other type of transaction behaviours.
See my PR #18430 , which has the implementation I am talking about.

**Comment by brainopia** [comment]
@arthurnn You are mistaken. Only transactions with a parent where `commit_records/participates_in_records_transaction = false` would commit records themselves. And only in fixtures in case of `use_transactional_fixtures = true` would a transaction with such setting be created.

And since your PR makes fixtures dependent on a lot of inner structure of Transactions, I think this PR is a finer solution. @jeremy what do you think?

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 5]
```diff
@@ -166,7 +166,14 @@ def begin_transaction(options = {})
       end
 
       def commit_transaction
-        @stack.pop.commit
+        transaction = @stack.pop
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: I think this would be several times less confusing if it was named `inner_transaction` or similar; having `transaction` and `current_transaction` be different things is nigh unreadable to me.

**Comment by matthewd** [comment]
@arthurnn @jeremy this more closely resembles my idea of the plan here -- that we were going to mark the outermost transaction as "please pretend not to be here", thus causing the next level down to adopt the behaviours of a non-savepoint transaction (.. apart from which commands get sent to the DB).

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 5 above)
Comment (brainopia) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: @matthewd indeed, thank you. I've reflected your note in code.

**Comment by arthurnn** [comment]
@brainopia you are right, this PR works indeed, sorry about that.
I got into that wrong conclusion, because I found a bit hard to follow how the commit_records happen in this PR, because it is always up to the previous transaction to decide if the next should `commit_records` or not.

I think the actual Transaction should have a flag, which will decide if should `commit_records` or not, see my PR https://github.com/rails/rails/pull/18438 , I cherry-pick your work/tests and added you name to the Changelog too..

**Comment by arthurnn** [comment]
Let me start this over, without actually writing code..
@brainopia what do you think about implementing something like a top level transaction? So RealTransaction would be a top-level by default, and Savepoint would not. And before we push the transaction to the stack, we can opt-out the top-level attribute, so the next transaction would be the top-level. 
And like that, the current transaction can decide if it needs to commit the records, based on the top-level flag.

**Comment by brainopia** [comment]
@arthurnn these are two sides of the coin.
In both cases parent transaction is checked either on creation of the nested transaction or commit of the nested transaction.

I'd prefer late binding that is checking the property of parent transaction during commit time It would allow a setting be more flexible and changed at any time. I'll demonstrate it with example:
Two transactions were created by default, one is parent and another is child. If afterwards we'll change the property `top_level/participates_in_records_transaction/commit_records` of the parent transaction, then the child can adopt the correct behavior if it would look up the property at commit time otherwise we'd need to change the property manually for both transactions.

Also the approach taken in #18438 was hardwired to current transaction types and this one does not depend upon them. And if someone adds a new one in a gem, it will just work with `top_level/participates_in_records_transaction/commit_records` without any work from developer.

And the last nitpick #18438 prevents ability to setup a nested transaction with `top_level/participates_in_records_transaction/commit_records`. It was made on purpose, but I'm not sure this constraint is needed.

The harder question is how to pick a name :smile: 
There are already three candidates: `top_level/participates_in_records_transaction/commit_records`

**Comment by brainopia** [comment]
Scratch my previous comment, lets reiterate.

The initial use-case was to run `after_commit` callbacks for transactions directly underneath a fixture transaction. That is to achieve

```ruby
connection.begin_transaction do # <= fixture transaction
  connection.begin_transaction do
    # ...
  end # <= commit callbacks happen here

  connection.begin_transaction do
    # ...
  end # <= commit callbacks happen here
end
```

There are two ways to influence how nested transactions are created:

- directly by using a special `TransactionManager`. Implementation of such approach in #18430 have required to rewrite part of internals outside their scope. And while it can be improved the `TransactionManager` is mostly an internal detail and it's easier from outside to work in terms of transactions.

- indirectly by changing an environment in which transactions are run to account whether `after_commit` callbacks should be fired. A common denominator for nested transactions is a parent transaction and we have control over it in our use-case. Therefore we'd like the code like following to work.

```ruby
connection.begin_transaction commit_callbacks: false do
  connection.begin_transaction do
    # ...
  end # <= commit callbacks happen here

  connection.begin_transaction do
    # ...
  end # <= commit callbacks happen here
end
```

If this is all we care about then the approach in current PR is the simplest one (it does not depend on types of involved transactions contrary to #18438). If this is the case we'd need only to pick a better name to show that only nested transactions are affected and not the transaction itself.

But ability to turn off `after_commit` callbacks for any transaction (without manipulating its parent transaction, which it does not have access to anyway) sounds interesting as well (even though I don't have a use-case yet). And since this is conceptually related to disabling `after_commit` callbacks for children it would be wise to implement it together and at once.

**Comment by arthurnn** [comment]
> If this is all we care about then the approach in current PR is the simplest one (it does not depend on types of involved transactions contrary to #18438)

Actually thats not true. 18438 doesnt depend on types involved, the code `transaction.top_level = true unless @stack.detect { |t| t.top_level }` is on the `SavepointTransaction` code path, but just as an optimization, that could be easily outside the if, and work for both Transactions.

> But ability to turn off `after_commit` callbacks for any transaction (without manipulating its parent transaction, which it does not have access to anyway)

I agree, I dont like having to touch the parent to change a behaviour on the current transaction (i.e. to  #commit_records)

Pretty much the big difference from my PR and ours is that I changed the name to  `top_level`, and once you think about it, `top_level` means the actual transaction that is the top_level that will call records callbacks, and not the parent. So I also changed the flag to live in the current transaction and not in the parent.
My main concern about this PR, is having to touch the parent on the commit code path (https://github.com/rails/rails/pull/18415/files#diff-52c8dd8e01c039c37b33b3fcbdfe406aR172) is a bad idea, if we can do it before pushing to the stack, and just read a boolean flag. 

Thoughts?

**Comment by arthurnn** [comment]
So if you wanna improve my PR, to improve this concern
> it does not depend on types of involved transactions

**Comment by arthurnn** [comment]
AHHHH :cry: (I hit the wrong button )

**Comment by brainopia** [comment]
@arthurnn no worries :smile: 

https://github.com/brainopia/rails/compare/fix_after_commit_for_fixtures...fix2_after_commit_for_fixtures - I've prepared a comparison with `top_level` implementation. I've changed a couple of things to make sure that at most one top_level transaction exists at any moment.

Do you still prefer this approach?
I'd like the opinions of others as well.

Also I want to bring up the question of `:joinable` setting. The only two places where it's used in rails are fixtures and console/sandbox. Fixtures will use `top_level: false` and console/sandbox should probably as well. There are no other uses in rails codebase or in any other publicly available code on github. Since `top_level: false` always needs `joinable: false` and I can't find a use-case for `joinable: false` without `top_level: false` then shouldn't we deprecate `:joinable` and just leave `:top_level`?
